firedrake/assemble.py

```diff
@@ -20,7 +20,7 @@
 
 
 def assemble(f, tensor=None, bcs=None, form_compiler_parameters=None,
-             inverse=False, nest=None):
+             inverse=False, nest=None, matfree=False, extra_ctx={}):
```

Comment:
`mat_type` is the petsc name, hence familiar. Don't worry about zero or one forms.

The petsc name for one forms would be `vec_type`. We don't support the different formats, so don't worry about that.